### PR TITLE
Add BF16 and FP16 Support

### PR DESCRIPTION
**Pull request by mikecovlee**
Currently, m-lora uses FP32. However, for devices with abundant video memory, you can load FP16 without using quantization to reduce the accuracy loss and overhead caused by quantization; for devices with insufficient video memory, you can further save time in the training process by using FP16 and quantization.

### CODE REVIEW

File: generate.py [hunk 1]
```diff
@@ -7,12 +8,15 @@ def main(base_model: str,
          input: str = None,
          template: str = None,
          lora_weights: str = None,
+         half_precision: bool = False,
```
Comment (yezhem) [inline generate.py:11]: just use load_16bit for align

File: mlora/lora_liner.py [hunk 2]
```diff
@@ -29,11 +29,11 @@ def set_parameter(self, r: int, alpha: int, dropout: float):
         self.scaling_ = alpha / r
 
     def forward(self, data: torch.Tensor) -> torch.Tensor:
-        data_ = F.dropout(data, self.dropout_)
+        data_ = F.dropout(data.type_as(self.lora_a_), self.dropout_)
```
Comment (yezhem) [inline mlora/lora_liner.py:32]: use to function, dose type_as is in-place op???

File: mlora/model_llama.py [hunk 3]
```diff
@@ -340,22 +360,23 @@ def from_pretrained(path: str,
...
             llama_args.pad_token_id_ = -1
+        llama_args.dtype = llama_model.dtype
 
         llama_args.device = device
 
         model = LlamaModel(llama_args)
 
         embedding_weight = llama_model.model.embed_tokens.weight.to(
-            device=device).requires_grad_(False)
+            device=device).detach()
```
Comment (yezhem) [inline mlora/model_llama.py:370]: why use detach to copy tensor??
can add some comment ??

**Review by yezhem — CHANGES_REQUESTED** [review]

File: mlora/lora_liner.py (hunk 2 above)
Comment (mikecovlee) [reply mlora/lora_liner.py:32]: image

File: mlora/model_llama.py (hunk 3 above)
Comment (mikecovlee) [reply mlora/model_llama.py:370]: image

**Comment by merlintang** [comment]
do we have a config to open fp16 vs fp32?

File: mlora/lora_liner.py (hunk 2 above)
Comment (yezhem) [reply mlora/lora_liner.py:32]: https://discuss.pytorch.org/t/why-does-pytorch-needs-the-three-functions-to-type-and-type-as/20164/4 use this

File: mlora/model_llama.py [hunk 4]
```diff
@@ -320,12 +340,12 @@ def from_pretrained(path: str,
                     bnb_4bit_use_double_quant=double_quant,
                     bnb_4bit_quant_type=quant_type,
                 ),
-                torch_dtype=(torch.float32 if fp16 else (torch.bfloat16 if bf16 else torch.float32)))
+                torch_dtype=dtype)
```
Comment (yezhem) [inline mlora/model_llama.py:343]: use dtype be argument name is dangerous

File: mlora/model_llama.py [hunk 5]
```diff
@@ -369,14 +390,14 @@ def from_pretrained(path: str,
                 layer.self_attn.o_proj, device=device)
             model.layers_[idx].ffn_ = FeedForward(
                 norm=RMSNorm(layer.post_attention_layernorm.weight.to(
-                    device=device).requires_grad_(False), model.norm_eps_),
+                    device=device).detach(), model.norm_eps_),
```
Comment (yezhem) [inline mlora/model_llama.py:393]: do not use detach function, because we need the tensor in calc graph, but does not need grad.

**Review by yezhem — CHANGES_REQUESTED** [review]

File: mlora/model_chatglm.py [hunk 6]
```diff
@@ -184,17 +184,32 @@ def forward_for_checkpoint(*inputs):
     def from_pretrained(path: str,
                         device: str,
                         bits: int = None,
-                        fp16: bool = True,
-                        bf16: bool = True,
+                        load_dtype: torch.dtype = torch.bfloat16,
```
Comment (yezhem) [inline mlora/model_chatglm.py:187]: maybe we should use the base_model_dtype and adapter_model_dtype??

**Review by yezhem — COMMENTED** [review]
LGTM

**Review by yezhem — APPROVED** [review]
LGTM
